packager/VisibilityChecker.cc

```diff
@@ -40,8 +40,14 @@ class PropagateVisibility final {
     }
 
     void recursiveExportSymbol(core::GlobalState &gs, bool firstSymbol, core::ClassOrModuleRef klass) {
-        // We only mark symbols from this package.
-        if (!this->definedByThisPackage(gs, klass)) {
+        // We only mark symbols from this package. However, there's a
+        // tough case where non-behavior-defining "namespace-like"
+        // constants might get attributed to other packages (since we
+        // don't have a canonical location, so we use the first place
+        // we see them... which might be in a subpackage) and
+        // therefore this might stop too soon. That's why we only stop
+        // recursing if the thing is actually behavior-defining.
+        if (!this->definedByThisPackage(gs, klass) && klass.data(gs)->flags.isBehaviorDefining) {
```

Comment:
I remember aadi saying something that he thought the isBehaviorDefining flag was not always correctly populated on the fast path.

I also see this comment:

https://github.com/sorbet/sorbet/blob/f4e43c95ebb97a03ed2b9d21c041a99f3901ee6f/namer/namer.cc#L230-L231

If we're starting to use this flag more in error messages we might want to make sure we have good fast path tests for these things. For example, we should check that making a fast path edit that either makes a class start defining behavior or stop defining behavior updates that bit on the class correctly.

Right now, I don't see anywhere that tells me that the `definesBehavior` flag is ever _unset_ if a class stops defining behavior.

Reply:
I think more tests is a good idea, but it's worth noting we're already using this flag in the visibility checker and I believe we have been since it was first introduced, which means we haven't had risen-to-the-level-of-visibility problems with it in that whole time. That doesn't mean we shouldn't introducer more strict tests and make sure it works in fast-path contexts—we should—but rather that problems with it seem to be relatively esoteric and we've got some wiggle room with it.